### DISCUSSION TITLE
Asyncio task creation and cancellation problem

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#16253] [Using MicroPython] [answered]
Trying to get two blinking LEDs to start: (corrected task1 and task2)

```py
async def test():
    while True:
        if push_button is True:
            task1 = asyncio.create_task(blink_red(house_red))
            task2 = asyncio.create_task(blink_green(gate_green))

asyncio.run(test())
```
sometimes didn't work.  So, I thought I would try to check  for proper
operation following this example:

```py
# SuperFastPython.com
# example of canceling a scheduled task
import asyncio
 
# define a coroutine for a task
async def task_coroutine():
    # report a message
    print('executing the task')
    # block for a moment
    await asyncio.sleep(1)
 
# custom coroutine
async def main():
    # report a message
    print('main coroutine started')
    # create and schedule the task
    task = asyncio.create_task(task_coroutine())
    # cancel the task before it has had a chance to run
    was_cancelled = task.cancel()
    # report whether the cancel request was successful
    print(f'was canceled: {was_cancelled}')
    # wait a moment
    await asyncio.sleep(0.1)
    # check the status of the task
    print(f'canceled: {task.cancelled()}')
    # report a final message
    print('main coroutine done')
 
# start the asyncio program
asyncio.run(main())
```
Running this program throws-up this error:
```
>>> 
MPY: soft reboot
main coroutine started
was canceled: True
Traceback (most recent call last):
  File "main.py", line 30, in <module>
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in run_until_complete
  File "main.py", line 25, in main
AttributeError: 'Task' object has no attribute 'cancelled'
MicroPython v1.23.0 on 2024-06-02; Generic ESP32 module with ESP32
Type "help()" for more information.
>>>
```
`task.cancel()` seemed to work, so can I assume that it really has been cancelled?  

Do you need to test that the task was really created, in the first place?

I want a system that is totally reliable.

P.S. while loops need `await async.sleep(0)` in them.

**Comment by beetlegigg** [comment, marked as answer]
The task class in Python has the cancelled method but in micropython it does not, and has a much smaller set of methods.  Have a look by adding print(dir(task)) to the Superfast code snippet and then run it in normal python and then run the code again in micropython, but deleting the print(f'canceled: {task.cancelled()}') line which will cause micropython to hiccup.  You can then see the python / micropython differences for async tasks

In your own code snippet I think the likes of task1 = asyncio.run(blink_red(house_red)) looks wrong.  I'm no expert, but I think you should only run one asyncio.run loop (i.e the async def test() function in your snippet). I've only created and run tasks with mytask = asyncio.create_task(somthing()) type of syntax.

**Reply by davefes** [reply]
Thanks for the explanation.  I should check with Python.

I have seen multiple tasks in some examples, so thought it would be OK.

Will try some switch debouncing.

**Reply by davefes** [reply]
When I added `gate_sensor = Pin('B15', Pin.OUT, value=0)` to the script below I got an error:
`NameError: local variable referenced before assignment'

Oops, I then added `global gate_sensor'
and the error disappeared and things seem to work better.  A few hours of testing might confirm proper behaviour.

```py
async def monitor():
    global gate_sensor
    
        
    while True:

        while gate_sensor.value() != 0:
            await asyncio.sleep_ms(DEBOUNCE)
            
            if gate_sensor.value() == 0:
                task1 = asyncio.create_task(blink_red(house_red))
                task2 = asyncio.create_task(blink_green(gate_green))
 
                # when finished make the GPIO output low
                gate_sensor = Pin('B15', Pin.OUT, value=0)

                break       

            print('looping') 
            await asyncio.sleep_ms(10)

        print('running')
        await asyncio.sleep_ms(5000)
```
Why didn't it complain before?

P.S. `house_red` and `gate_green` should probably be `global` as well.

**Reply by beetlegigg** [reply]
Its to do with creating a local variable with the same name. In the code snippet below the function will happily reference the global myglob and print it out.  But as soon as you then put myglob = whatever in the function then in the whole of the function myglob has local scope and whereas it would previously happily print myglob, now it will only look for that name in the local scope and will tell you off for trying to print myglob before its assigned.
```
myglob = 2

def myfunc():
    print(myglob)
    # if the following line is uncommented then the print above will error
    #myglob = 3
    
myfunc()
```
so as soon as you put gate_sensor = Pin('B15', Pin.OUT, value=0) in your function then gate_sensor is in local scope for the whole function unless its declared as global.
Im not sure I have explained this very well, but I hope you understand what I mean :)

**Reply by davefes** [reply]
Thanks again. I understand your explanation ... but I would have thought that `while gate_sensor.value() != 0:` and `if gate_sensor.value() == 0:` without `global gate_sensor` would have complained.whereas adding:
`gate_sensor = Pin('B15', Pin.OUT, value=0)`
it did.

Maybe, it was "complaining in the background".

**Reply by davefes** [reply]
Seems testing is OK but assignment is not.

**Reply by beetlegigg** [reply]
The while gate_sensor.value !=0: is just reading the variable, and if it cant find this name in the local scope it will happily then look in the global scope.  
But as soon as you made the assignment  in the function with  'gate_sensor = xxx' then the local function does now have a local name so it will no longer look to the global scope.  And, as the function is asked to test 'while gate_sensor.value !=0' before the gate_sensor is assigned further down your code, then it feels the need to loudly complain :).   So it was not complaining in the background but was quite a happy bunny until you make the assignment and got python all grumpy.

**Reply by davefes** [reply]
OK, now I get it.  I don't think I was fully-aware that it will go up to global scope, if it doesn't find a local name.  In my many scripts I must have declared variables as `global` in supporting methods, if needed ... but didn't understand the scope of my "oops".

Appreciated the help.

**Comment by peterhinch** [comment]
@beetlegigg is correct with

> In your own code snippet I think the likes of task1 = asyncio.run(blink_red(house_red)) looks wrong. I'm no expert, but I think you should only run one asyncio.run loop (i.e the async def test() function in your snippet). I've only created and run tasks with mytask = asyncio.create_task(somthing()) type of syntax.

In typical applications the call to `asyncio.run()` never returns: it is used to start the main application task. `.run()` only returns if the main application task terminates or under error conditions. See [the tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#21-program-structure).

**Reply by davefes** [reply]
@peterhinch Thanks for this comment.  I now see that I had made mistakes in posting the original snippet.  The two tasks were actually:
```py
task1 = asyncio.create_task(blink_red(house_red))
task2 = asyncio.create_task(blink_green(gate_green))

**Reply by davefes** [reply]
Just realised that there is a bigger problem with my initial code snippet.  I found the putting a short delay in the while loop got the LEDs blinking.   Then I found:
https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#9-polling-vs-interrupts
specifically 8.4 Scheduling in asyncio
where `await asyncio.sleep(0)` are needed.
